### PR TITLE
shell: Fix browser history

### PR DESCRIPTION
**Pull request by mvollmer**
The browser creates an entry in the history whenever an iframe changes
its "window.location" (which is how our pages implement
in-page-navigation).  The shell did not expect this and would add
another entry whenever the location of the current iframe changes.

---

### The back and forward buttons in your browser now work as expected

Cockpit used to interfere with the browser history, and hitting the back button would frequently get into a loop, for example. This has been fixed.

### CODE REVIEW

**Comment by mvollmer** [comment]
Looks like this broke `cockpit.jump` somehow...

**Comment by mvollmer** [comment]
> Looks like this broke `cockpit.jump` somehow...

No, tracking the hash needs to call `navigate` to get all the internal state updated.  `jump(..., false) would not do that, but `navigate` will call `jump`, so let's call `navigate` directly.

**Comment by mvollmer** [comment]
I'll add a test.

**Review by garrett — APPROVED** [review]
Wow! This is already _so much better!_

It's not perfect, and there are some issues across some pages, but this feels like it's 99% solved.

Example I found where there's an issue:
- Navigate to accounts, then user, then back to accounts and try the back/forward buttons there.

But, provided the code and tests are good, I think this should probably be merged. The account issue is probably something else, as would be anything else that might be currently unknown.

We might even want to mention this fix in the release notes as it does impact user experience quite a bit and this was a longstanding bug. WDYT?

**Review by garrett — CHANGES_REQUESTED** [review]
Please look into the issues I mentioned above. Perhaps they're outliers. If so, then I'll re-approve this. Otherwise, if they do affect this PR, then I'll be happy to review again later.

**Comment by mvollmer** [comment]
>     1. Navigate to accounts, 
>     2. then a user,
>     3. then back to the accounts list

This means clicking the "Accounts" link in the bread crumbs, right?  This does indeed put two entries into the history instead of the expected single one.

**Comment by mvollmer** [comment]
> This means clicking the "Accounts" link in the bread crumbs, right? This does indeed put two entries into the history instead of the expected single one.

All of our breadcrumbs put two entries into the history, even when not embedded into the shell, because of how they are implemented.

```
<BreadcrumbItem onClick={() => cockpit.location.go("/")} to="#">{_("Accounts")}</BreadcrumbItem>
```

This turns into

```
<li onClick=...>
  <a href="#">...</a>
</li>
```

Both the `a` and `li` will cause the browser to add a history entry. But this is not the whole story...

**Comment by mvollmer** [comment]
> Both the `a` and `li` will cause the browser to add a history entry. But this is not the whole story...

We can use this:

```
 <BreadcrumbItem to="#/">{_("Accounts")}</BreadcrumbItem>
```

This will cause the browser to only add one entry.  Using `#/` instead of `#` is important:

- The shell will normalize a empty hash to "#/" (but only for the "src" attribute of the iframe)
- So when the iframe itself changes its hash to `#`, the shell will set it's `src` attribute to end in `#/`.
- This counts as another navigation event and the browser adds one more entry to the history.

**Comment by mvollmer** [comment]
> Both the `a` and `li` will cause the browser to add a history entry.

And they only do that because they navigate to *different* locations, the `li` goes to "...#/" and the `a` goes to "...#".  So this also works:

```
<BreadcrumbItem onClick={() => cockpit.location.go("/")} to="#/">{_("Accounts")}</BreadcrumbItem>
```

but the onClick really shouldn't be necessary.

**Comment by mvollmer** [comment]
> 2\. Click a service under CPU or Memory

This would two entries because this was also a link with both a click handler and a href="#".

**Comment by mvollmer** [comment]
Another piece of the puzzle was that cockpit.jump with a hash would add two entries because the hash change alone would cause the browser to add one, and then the jump command would add another.

**Comment by mvollmer** [comment]
> Another piece of the puzzle was that cockpit.jump with a hash would add two entries because the hash change alone would cause the browser to add one, and then the jump command would add another.

Unfortunately, the history that is added by the browser does not create a "popstate" event when removed from the history stack, the browser just undoes the hash change. Thus the frame change doesn't happen and things get confusing.

Hmm....

**Comment by mvollmer** [comment]
> Unfortunately, the history that is added by the browser does not create a "popstate" event when removed from the history stack, the browser just undoes the hash change. Thus the frame change doesn't happen and things get confusing.

We can avoid the extra history entry from setting the "src" attribute by forcing the window.location to the right value beforehand, with window.location.replace.

**Review by garrett — APPROVED** [review]
I've been running the current version of this PR across browsers (Firefox, Chrome, GNOME Web, & Safari iOS) and haven't found an issue.

Congrats! I think this is solved! If there's anything not quite working, it would be a corner case — and I haven't found it in my latest round of random testing on various pages on various browsers.

The only one small "gotcha" is that the scroll state doesn't seem to be preserved. That is, scroll down a page, navigate to a new one, and hit back. The page should be at the same location as where you left it. (Same for going forward.) Browsers do this by default and it should be our behavior here too. It seems that something in the code has the pages jump to the top when it shouldn't.

However, I think this is a _HUGE_, dramatic improve over what we've had for several years, so I'm excited for this to go in ASAP. Well done! I think we can investigate the scroll issue as a follow-up.

**Comment by garrett** [comment]
The History.scrollRestoration might need to be set:

- https://developer.mozilla.org/en-US/docs/Web/API/History/scrollRestoration
- https://html.spec.whatwg.org/multipage/history.html#she-scroll-restoration-mode

It's supposed to be set at the top of the document. (I wonder if we set it to manual somewhere.)

**Comment by garrett** [comment]
I've looked through Cockpit code and don't see any obvious setting of history.scrollRestoration or scrollTo / scrollTop (except in updates.js, which wouldn't affect everything).

However, when grepping through dist, scrollTo / scrollTop shows up a LOT. (This was just a quick look.) I think it might be somewhere in React or PatternFly. It's all the more reason to land this PR and _then_ look into the scroll history stuff... it'll probably take some debugging and possibly messing with our stack. :cry: 

Meanwhile, what's in this PR is a 10000% improvement. :+1:

**Comment by mvollmer** [comment]
> I've been running the current version of this PR across browsers (Firefox, Chrome, GNOME Web, & Safari iOS) and haven't found an issue.

YES! :-) Thanks a lot!

**Comment by mvollmer** [comment]
> The only one small "gotcha" is that the scroll state doesn't seem to be preserved. 

Right, let's look at this in a separate PR.

File: pkg/users/account-details.js [hunk 1]
```diff
@@ -226,7 +226,7 @@ export function AccountDetails({ accounts, groups, shadow, current_user, user })
                 </Title>
                 <EmptyStateSecondaryActions>
                     <Breadcrumb>
-                        <BreadcrumbItem onClick={() => cockpit.location.go("/")} to="#">{_("Back to accounts")}</BreadcrumbItem>
```
Comment (marusak) [inline pkg/users/account-details.js:229]: Do we need to fix this in other projects as well? If they don't fix it, does it break them or just the history is gonna be broken as before?

**Review by marusak — APPROVED** [review]
Works perfectly, thanks!

File: pkg/users/account-details.js (hunk 1 above)
Comment (mvollmer) [reply pkg/users/account-details.js:229]: Just the history will be broken. We should check for this pattern also in the external projects.

**Comment by garrett** [comment]
I think this is a sizable enough UX improvement to merit a mention in the release notes. Yes, it's a bug fix, but it's a very important one. Plus, we should have an update with Cockpit Client, which will have a visible UI change (addition of back/forward buttons). Both can be paired together in the final release notes.

File: pkg/users/account-details.js (hunk 1 above)
Comment (mvollmer) [reply pkg/users/account-details.js:229]: I have checked c-machines, c-podman, and starter-kit. None of them uses this pattern.

**Comment by mvollmer** [comment]
testHistory failure on coreos is not a flake, investigating...

**Comment by mvollmer** [comment]
> testHistory failure on coreos is not a flake, investigating...

CoreOS does not seem to keep login history, so that part of the test simply needs to be skipped.

**Comment by mvollmer** [comment]
@garrett, @marusak, the only changes since your last review were to get the test pass on CoreOS. No changes to Cockpit code.

**Review by garrett — APPROVED** [review]
Re-approving, as only tests were changed.

I'm looking forward to this landing!

**Comment by martinpitt** [comment]
Let's land PR #16987 first to fix the fedora-coreos woes, and then retry.

**Pull request merged**